llama_index/schema.py

```diff
@@ -21,20 +23,66 @@ def _validate_is_flat_dict(metadata_dict: dict) -> None:
             raise ValueError("Value must be one of (str, int, float, None)")
 
 
-@dataclass
-class BaseDocument(DataClassJsonMixin):
-    """Base document.
+class NodeRelationship(str, Enum):
+    """Node relationships used in `BaseNode` class.
 
-    Generic abstract interfaces that captures both index structs
-    as well as documents.
+    Attributes:
+        SOURCE: The node is the source document.
+        PREVIOUS: The node is the previous node in the document.
+        NEXT: The node is the next node in the document.
+        PARENT: The node is the parent node in the document.
+        CHILD: The node is a child node in the document.
 
     """
 
-    # TODO: consolidate fields from Document/IndexStruct into base class
-    text: Optional[str] = None
-    doc_id: Optional[str] = None
-    embedding: Optional[List[float]] = None
-    doc_hash: Optional[str] = None
+    SOURCE = auto()
+    PREVIOUS = auto()
+    NEXT = auto()
+    PARENT = auto()
+    CHILD = auto()
+
+
+class ObjectType(str, Enum):
+    DOCUMENT = auto()
+    TEXT = auto()
+    IMAGE = auto()
+    INDEX = auto()
+
+
+class MetadataMode(str, Enum):
+    ALL = auto()
+    EMBED = auto()
+    LLM = auto()
+    NONE = auto()
+
+
+class RelatedNodeInfo(BaseModel):
+    node_id: str
+    node_type: Optional[ObjectType] = None
+    metadata: Dict[str, Any] = Field(default_factory=dict)
+    hash: Optional[str] = None
+
+
+RelatedNodeType = Union[RelatedNodeInfo, List[RelatedNodeInfo]]
+
+
+class BaseNode(BaseModel):
+    """Base node Object.
+
+    Generic abstract interface for retrievable nodes
+
+    """
+
+    id_: str = Field(
+        default_factory=lambda: str(uuid.uuid4()), description="Unique ID of the node."
+    )
+    embedding: Optional[List[float]] = Field(
+        default=None, description="Embedding of the node."
+    )
+    weight: float = Field(
```

Comment:
would the weight change during query-time? or is it fixed per node? trying to think about what this property represents / if we should put in NodeWithScore instead

Reply:
It's fixed per-node. I.e. when you create a node/document, you can give a weight during index creation

Reply:
Hope that makes sense!

Reply:
I also still feel a bit hesitant on this. Could we defer adding this until we have the proper support on the retrieval/ranking side? Feels a bit speculative right now.

Reply:
I did add support to our simple vector store actually. I can remove aspect if it's an issue then 👍🏻 

The dimensions in vectara are essentially our metadata no? Since the metadata text contributes to the embeddings?

Reply:
RE support in simple vector store: ah okay, guessing it only works for the simple vector store and not others right now? 

Also took a quick look, seems like we just multiply the weight with the score? probably should document this an give guidance on how to set the weight (doesn't have to be this PR).

RE vectara dimensions: not the same as metadata, it's used to adjust resulting score dynamically by query: https://docs.vectara.com/docs/common-use-cases/semantic-search/custom-dimensions

I kinda see it as a generalization of the `weight` here. Since it supports:
1.  multiple "dimensions" for score adjustment, rather than just a single weight
2. allow per-query adjustment to score

but anyhow, this is probably out of scope here.

Reply:
Yea I did add a doc page for it, but meh, might remove for now.

Also interesting point on vectara. Missed that page of their docs